ext/pdo_pgsql/pgsql_statement.c

```diff
@@ -513,6 +517,18 @@ static int pgsql_stmt_get_col(pdo_stmt_t *stmt, int colno, zval *result, enum pd
 #endif
 				ZVAL_LONG(result, ZEND_ATOL(ptr));
 				break;
+			case FLOAT4OID:
+			case FLOAT8OID:
+                if (len == sizeof("Infinity") - 1 && strcmp(ptr, "Infinity") == 0) {
```

Comment:
why not doing strncmp directly ?

Reply:
strncmp is pretty optimised :) I m not sure about the trade off readibility/theorical performance increase.

Reply:
I think it is necessary to optimize.

My test code:

```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/time.h>

double tod1(const char *ptr, size_t len)
{
    if (len == sizeof("Infinity") - 1 && strncmp(ptr, "Infinity", len) == 0) {
        return -1;
    } else if (len == sizeof("-Infinity") - 1 && strncmp(ptr, "-Infinity", len) == 0) {
        return -1;
    } else if (len == sizeof("NaN") - 1 && strncmp(ptr, "NaN", len) == 0) {
        return -1;
    } else {
        strtod(ptr, NULL);
    }
}

double tod2(const char *ptr, size_t len)
{
    if (strncmp(ptr, "Infinity", len) == 0) {
        return -1;
    } else if (strncmp(ptr, "-Infinity", len) == 0) {
        return -1;
    } else if (strncmp(ptr, "NaN", len) == 0) {
        return -1;
    } else {
        strtod(ptr, NULL);
    }
}

const char* STRS[] = {
    "1",
    "1.2",
    "3.14",
    "3.14",
    "123.456"
};

void test1() {
    for (size_t i = 0; i < sizeof(STRS) / sizeof(STRS[0]); ++i) {
        struct timeval start_time, end_time;

        const char *str = STRS[i];
        size_t len = strlen(str);

        gettimeofday(&start_time, NULL);

        for (size_t j = 0; j < 1000000; ++j) {
            tod1(str, len);
        }

        gettimeofday(&end_time, NULL);

        double execution_time = (double)(end_time.tv_sec - start_time.tv_sec) + (double)(end_time.tv_usec - start_time.tv_usec) / 1000000.0;

        printf("test with length check \"%s\" time: %f ms\n", STRS[i], execution_time * 1000);
    }
}

void test2() {
    for (size_t i = 0; i < sizeof(STRS) / sizeof(STRS[0]); ++i) {
        struct timeval start_time, end_time;

        const char *str = STRS[i];
        size_t len = strlen(str);

        gettimeofday(&start_time, NULL);

        for (size_t j = 0; j < 1000000; ++j) {
            tod2(str, len);
        }

        gettimeofday(&end_time, NULL);

        double execution_time = (double)(end_time.tv_sec - start_time.tv_sec) + (double)(end_time.tv_usec - start_time.tv_usec) / 1000000.0;

        printf("test no length check \"%s\" time: %f ms\n", STRS[i], execution_time * 1000);
    }
}

int main() {
    test1();
    test2();
    return 0;
}
```

output:

```log
# gcc -O3 -o test test.c && ./test
test with length check "1" time: 19.308000 ms
test with length check "1.2" time: 34.546000 ms
test with length check "3.14" time: 37.638000 ms
test with length check "3.14" time: 35.310000 ms
test with length check "123.456" time: 44.804000 ms
test no length check "1" time: 62.552000 ms
test no length check "1.2" time: 79.284000 ms
test no length check "3.14" time: 79.511000 ms
test no length check "3.14" time: 80.326000 ms
test no length check "123.456" time: 90.000000 ms
```

Reply:
Alright I get this
`./a.out
test with length check "1" time: 22.389000 ms
test with length check "1.2" time: 25.469000 ms
test with length check "3.14" time: 18.486000 ms
test with length check "3.14" time: 15.308000 ms
test with length check "123.456" time: 14.562000 ms
test no length check "1" time: 12.050000 ms
test no length check "1.2" time: 15.771000 ms
test no length check "3.14" time: 16.043000 ms
test no length check "3.14" time: 15.968000 ms
test no length check "123.456" time: 16.557000 ms`

let's keep it then you were right :)

Reply:
By the way the same test on Linux/glibc (previous was on macOs)

`./a.out 
test with length check "1" time: 50.399000 ms
test with length check "1.2" time: 33.425000 ms
test with length check "3.14" time: 35.650000 ms
test with length check "3.14" time: 35.656000 ms
test with length check "123.456" time: 44.160000 ms
test no length check "1" time: 25.461000 ms
test no length check "1.2" time: 40.037000 ms
test no length check "3.14" time: 41.458000 ms
test no length check "3.14" time: 41.551000 ms
test no length check "123.456" time: 50.772000 ms`

That might depend on the arch/libc version I guess.

Reply:
I trust your judgement about not just using zend_strtod below, however would like to remove the length check.